cirq-core/cirq/transformers/dynamical_decoupling.py

```diff
@@ -76,134 +75,122 @@ def _validate_dd_sequence(dd_sequence: Tuple['cirq.Gate', ...]) -> None:
         )
 
 
-def _parse_dd_sequence(schema: Union[str, Tuple['cirq.Gate', ...]]) -> Tuple['cirq.Gate', ...]:
-    """Parses and returns dynamical decoupling sequence from schema."""
+def _parse_dd_sequence(
+    schema: Union[str, Tuple['cirq.Gate', ...]]
+) -> Tuple[Tuple['cirq.Gate', ...], Dict['cirq.Gate', 'cirq.Pauli']]:
+    """Parses and returns dynamical decoupling sequence and its associated pauli map from schema."""
+    dd_sequence = None
     if isinstance(schema, str):
-        return _get_dd_sequence_from_schema_name(schema)
+        dd_sequence = _get_dd_sequence_from_schema_name(schema)
     else:
         _validate_dd_sequence(schema)
-        return schema
+        dd_sequence = schema
+
+    # Map Gate to Puali gate. This is necessary as dd sequence might contain gates like X^-1.
+    pauli_map: Dict['cirq.Gate', 'cirq.Pauli'] = {}
+    for gate in dd_sequence:
+        pauli_gate = _pauli_up_to_global_phase(gate)
+        if pauli_gate is not None:
+            pauli_map[gate] = pauli_gate
+    for gate in [cirq.X, cirq.Y, cirq.Z]:
+        pauli_map[gate] = gate
+
+    return (dd_sequence, pauli_map)
 
 
 def _is_single_qubit_operation(operation: 'cirq.Operation') -> bool:
-    if len(operation.qubits) != 1:
-        return False
-    return True
+    return len(operation.qubits) == 1
 
 
 def _is_single_qubit_gate_moment(moment: 'cirq.Moment') -> bool:
-    for operation in moment:
-        if not _is_single_qubit_operation(operation):
-            return False
-    return True
+    return all([_is_single_qubit_operation(op) for op in moment])
```

Comment:
a generator is more efficient here. the old code will build the full list before checking.

```suggestion
    return all(_is_single_qubit_operation(op) for op in moment)
```